Update to Infinispan v14 and use the reactive MongoDB driver

antoniomacri opened the pull request:
Fixes #22

tristantarrant commented on src/test/java/org/infinispan/persistence/mongodb/config/MongoDbStoreConfigurationSerializerTest.java:
```diff
@@ -0,0 +1,8 @@
+
+@Test(testName = "org.infinispan.persistence.mongodb.config.MongoDbStoreConfigurationSerializerTest", groups = "functional")
+public class MongoDbStoreConfigurationSerializerTest extends AbstractConfigurationSerializerTest {
+}
```
This class needs to override `compareStoreConfiguration()` like https://github.com/infinispan/infinispan/blob/main/persistence/rocksdb/src/test/java/org/infinispan/persistence/rocksdb/config/ConfigurationSerializerTest.java#L13

tristantarrant reviewed (changes requested):
We also need an XSD like https://github.com/infinispan/infinispan/blob/main/persistence/rocksdb/src/main/resources/schema/infinispan-cachestore-rocksdb-config-15.0.xsd

tristantarrant commented on src/main/java/org/infinispan/persistence/mongodb/config/ConnectionConfiguration.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+
+    private final Attribute<String> uri;
```
no need to have these attributes as fields, since they will be accessed infrequently. Just reference the attributes in the getters below

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+        Object storeKey = cacheToStoreConverter.toStoreKey(key);
+        Bson filter = and(eq("_id", storeKey), or(gt(EXPIRY_TIME, now), lte(EXPIRY_TIME, -1L)));
+
+        return Maybe.fromPublisher(collection.find(filter))
```
I personally would prefer something like `Flowable.fromPublisher(collection.find(filter)).singleStage()` to ensure we only ever receive a single entry from the Publisher, but I am fine with the way it is. TBH, I am surprised that `Maybe.fromPublisher` allows for more than 1 entry returned from a Publisher but `Single.fromPublisher` throws an exception.

And I know that since we are using _id that it shouldn't return more than one, but it is more protection of putting in a bug we don't expect.

antoniomacri replied:
From the doc, `Maybe.fromPublisher` should emit "the very first item", so (unless I'm misunderstanding something) it shouldn't be a problem.

Correct?

wburns replied:
That is fine, I was more talking about the case when the publisher publishes more than one item, but this is a minor thing that is fine if not changed.

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+        query.put("_id", cacheToStoreConverter.toStoreKey(key));
+
+        return Single.fromPublisher(collection.deleteOne(query))
+                .map(e -> e.getDeletedCount() > 0)
```
I am not familiar with the API, but do we need to check if the Result was acknowledged first?

antoniomacri replied:
You're right, fixed.

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+    }
+
+    @Override
+    public CompletionStage<Void> write(int segment, MarshallableEntry<? extends K, ? extends V> marshallableEntry) {
```
It looks like the `batch` command was not overridden. It looks like the mongo client is perfect for this method as well. Can we add that?

The `batch` method is used with ISPN transactions, batch API and when configured as an async store. Note that if the mongo client recommends having a bunch of smaller batch to us the `StoreConfiguration#maxBatchSize` configuration to determine how large of a batch to send to the mongo server.

wburns replied:
You can use the `bulkWrite` method where a batch is sent via using `Flowable.buffer(context.getConfiguration().maxBatchSize())` to create the batches. Just make sure to use concatMap when creating the sub Publishers so we only have one outstanding batch a time.

antoniomacri replied:
OK, I'm going to add the `batch` method.

antoniomacri replied:
Sorry, I'm not familiar with RxJava, and I cannot make it work with `concatMap` (instead of `concatMapEager`), unless I explicitly handle backpressure.

With this:
```java
        Flowable<WriteModel<Document>> entriesWritten = Flowable.fromPublisher(writePublisher)
                .concatMap(Flowable::fromPublisher)
                .map(e -> {
                    var storeKey = cacheToStoreConverter.toStoreKey(e.getKey());
                    return new ReplaceOneModel<>(eq("_id", storeKey), cacheToStoreConverter.toStoreEntry(storeKey, e), new ReplaceOptions().upsert(true));
                });

        Flowable<WriteModel<Document>> removedKeys = Flowable.fromPublisher(removePublisher)
                .concatMap(Flowable::fromPublisher)
                .map(key -> {
                    var storeKey = cacheToStoreConverter.toStoreKey(key);
                    return new DeleteOneModel<>(eq("_id", storeKey));
                });

        return Flowable.concatArrayEager(entriesWritten, removedKeys)
                .buffer(context.getConfiguration().maxBatchSize())
                .concatMap(writeModels -> Flowable.fromPublisher(collection.bulkWrite(writeModels)))
                .ignoreElements()
                .toCompletionStage(null);
```
`testWriteAndDeleteBatch` fails with:
> io.reactivex.rxjava3.exceptions.MissingBackpressureException: Unable to emit a new group (`#2`) due to lack of requests. Please make sure the downstream can always accept a new group as well as each group is consumed in order for the whole operator to be able to proceed.

Any ideas?

wburns replied:
Yeah, the concatMap here only subscribes to a single publisher at a time. You need to use `concatMapEager(Function, publisherCount, publisherCount)` on both the write and remove `Publisher<SegmentedPublisher>`s since there are multiples generated from the `groupBy` operator.

Note the default impl does something very similar on the interface as well.

antoniomacri replied:
Yes, sorry, I thought you meant to use exactly `concatMap` (not `*Eager`) in the two branches, since I've seen from the `batch` doc that only the `Flowable.concatArrayEager` is required for the two input publishers.

But indeed (now I had the time to look a bit deeper) also the groups require eager subscription.

Thanks

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+    @Override
+    public Publisher<MarshallableEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean includeValues) {
+        return Flowable.defer(() -> {
+            FindPublisher<Document> iterable = collection.find();
```
We may want to use the `StoreConfiguration#maxBatchSize` here for the `FindPublisher.batchSize`. I don't know how large of a batch the server would normally send back.

antoniomacri replied:
Added.

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+            }
+
+            return Flowable.fromPublisher(iterable)
+                    .filter(e -> filter == null || filter.test(cacheToStoreConverter.toCacheKey(e.get("_id"))))
```
Nitpick: I would do the inverse like
```java
Flowable<Document> flowable = Flowable.fromPublisher(iterable);
if (filter != null) {
   flowable = flowable.filter(e -> filter.test(cacheToStoreConverter.toCacheKey(e.get("_id"))));
}
return flowable..map(cacheToStoreConverter::toCacheEntry);
```

antoniomacri replied:
Done.

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+            }
+
+            return Flowable.fromPublisher(iterable)
+                    .filter(e -> filter == null || filter.test(cacheToStoreConverter.toCacheKey(e.get("_id"))))
```
Nitpick: I would do the inverse like
```java
Flowable<Document> flowable = Flowable.fromPublisher(iterable);
if (filter != null) {
   flowable = flowable.filter(e -> filter.test(cacheToStoreConverter.toCacheKey(e.get("_id"))));
}
return flowable..map(cacheToStoreConverter::toCacheEntry);
```

wburns commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+
+        return Flowable.fromPublisher(collection.find(filter))
+                .map(cacheToStoreConverter::toCacheEntry)
+                .concatWith(Flowable.defer(() -> Flowable.fromPublisher(collection.deleteMany(filter)).mapOptional(x -> Optional.empty())));
```
Unfortunately, this has a pretty big window where delete may not return the entries that were actually deleted.

Looking at the mongo API there isn't a great way to correlate a delete to a specific key. I wonder though if bulkWrite with a List of DeleteOneModel instances will return a Publisher with the same results if the option is ordered. If so then we can return a consistent view of what entries were actually removed due to expiration.

antoniomacri replied:
Something like this?
```java
        return Flowable.fromPublisher(collection.find(filter))
                .buffer(context.getConfiguration().maxBatchSize())
                .concatMap(list -> {
                    var deleteModels = list.stream()
                            .map(e -> new DeleteOneModel<Document>(eq("_id", e.get("_id"))))
                            .toList();
                    return Flowable.fromPublisher(collection.bulkWrite(deleteModels))
                            .mapOptional(ignored -> Optional.<MarshallableEntry<K, V>>empty())
                            .concatWith(Flowable.fromStream(list.stream().map(cacheToStoreConverter::toCacheEntry)));
                });
```

In `BulkWriteResult` upserts don't take into account deletes, so I don't see a way to get the ids/documents back.

wburns replied:
This is very close, we should also add the expiration time to the DeleteOneModel so it only deletes the entry if it is still expired. It is possible for a concurrent update that could change the expiration time.

I think the publishing of the entry based on if it was expired isn't quite right either

I believe the following should work, assuming the BulkWriteResult are returned in the same order as the write

```java
        return Flowable.fromPublisher(collection.find())
              .buffer(context.getConfiguration().maxBatchSize())
              .concatMap(list -> {
                  var deleteModels = list.stream()
                        .map(e -> new DeleteOneModel<Document>(eq("_id", e.get("_id"))), <expiration here>)
                        .toList();
                  return Flowable.zip(Flowable.fromPublisher(collection.bulkWrite(deleteModels)),
                        Flowable.fromIterable(list),
                        (bwr, entry) -> {
                            if (bwr.wasAcknowledged() && bwr.getDeletedCount() > 0) {
                                return Optional.of(cacheToStoreConverter.toCacheEntry(entry));
                            }
                            return Optional.<MarshallableEntry<K, V>>empty();
                        })
                        .mapOptional(Functions.identity());
              });
```

antoniomacri replied:
Unfortunately `bulkWrite` returns a publisher with a single result.

wburns replied:
That is quite disappointing.... Without that, we won't have proper expiration notifications for entries that are in the store but not in memory. The store will just need to be documented with that limitation.

antoniomacri replied:
What do you think about this?

```java
    public Publisher<MarshallableEntry<K, V>> purgeExpired() {
        long now = context.getTimeService().wallClockTime();

        return Flowable.fromPublisher(collection.find(and(lte(EXPIRY_TIME, now), gt(EXPIRY_TIME, -1L))).projection(Projections.include("_id")))
                .concatMapMaybe(d -> Single.fromPublisher(collection.findOneAndDelete(and(eq("_id", d.get("_id")), lte(EXPIRY_TIME, now), gt(EXPIRY_TIME, -1L))))
                        .mapOptional(docOrNull -> Optional.ofNullable(docOrNull).map(doc -> cacheToStoreConverter.toCacheEntry(doc))));
    }
```

Of course it's not the most efficient (without batching), but at least it doesn't iterate over the entire collection and should be safe.

Also it's not optimal, since it may miss some expired entries: is it a problem?

wburns replied:
I am fine either way. This should be fine the way it is, the entries can just be found on the next pass.

The only thing that worries me is that the `MongoCollection` API seems to be breaking reactive streams specification with `If no documents matched the query filter, then null will be returned`. A publisher cannot contain a null value, so I am not sure this will work or not...

wburns replied:
Sounds good.

wburns replied:
If you want to update to either do the more efficient removal but we don't have notifications or we can do the slower but have notifications. I guess I would lean towards the latter for correctness, but I don't know how much this will affect the mongo server with all those requests.

antoniomacri replied:
Same here.

antoniomacri commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+                .applyConnectionString(connString)
+                .retryWrites(true)
+                .build();
+        this.mongoClient = MongoClients.create(settings);
```
I don't know. I can investigate but the doc doesn't say anything, so maybe there is no guarantee.

Better to wrap in a `runBlocking`?

wburns replied:
Up to you, we may want to add block hound as optional like other modules do. That should catch it if is blocking.

https://github.com/infinispan/infinispan/blob/main/core/pom.xml#L89-L93

antoniomacri replied:
I used `blockhound` in a local test (checking for a false negative with a `Thread.sleep`) and it doesn't complain.

I'd leave as is.

antoniomacri commented on src/main/java/org/infinispan/persistence/mongodb/MongoDbStore.java:
```diff
@@ -0,0 +1,193 @@
+
+    @Override
+    public CompletionStage<Void> stop() {
+        mongoClient.close();
```
As above. Actually I remember of not being able to tell when the connection is actually closed, so maybe it's async.

`runBlocking`?

antoniomacri replied:
Same as above.

antoniomacri commented on src/main/java/org/infinispan/persistence/mongodb/converter/AbstractCacheToStoreConverter.java:
```diff
@@ -0,0 +1,78 @@
+    public Document toStoreEntry(Object storeKey, MarshallableEntry<? extends K, ? extends V> marshallableEntry) {
+        Document document = new Document("_id", storeKey)
+                .append(VALUE, toStoreValue(marshallableEntry.getValue()))
+                .append(METADATA, toByteArray(marshallableEntry.getMetadata()))
```
Added, but I'm wondering how it can be tested. From that I see, the base tests don't take it into account.

Any suggestion?

antoniomacri commented:
@wburns I have a question on the `publishEntries` method: if `!includeValues`, I'm actually including only keys.

Should I also include metadata/internalMetadata and the other attributes?

antoniomacri commented:
I implemented most of the requested changes (comments above on `batch` and `purgeExpired`).

I also fixed a bug in the parser and changed "converter=json" to "format=structured", which seems a better naming to me.

wburns reviewed (changes requested):
Sorry for the delay @antoniomacri, it is looking really good. Once the remaining couple changes are done I am fine integrating!

antoniomacri commented:
A few last questions:
1. How about the deployment? Do we need a github workflow?
2. Can I remove `.travis.yml`?

wburns commented:
> A few last questions:
> 
> 1. How about the deployment? Do we need a github workflow?

We can add this after this one is integrated.
> 2. Can I remove `.travis.yml`?

 I don't see why we can't just remove it.
 
 Once the expiration method is updated, I will integrate this PR, thanks again @antoniomacri !

antoniomacri commented:
@wburns `purgeExpired` updated.

Thank you!

wburns commented:
Squashed and merged, thanks @antoniomacri !
